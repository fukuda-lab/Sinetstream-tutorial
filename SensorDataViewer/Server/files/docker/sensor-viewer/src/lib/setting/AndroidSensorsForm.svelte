<script lang="ts" context="module">
  export const sensors = [
    "accelerometer",
    "accelerometer uncalibrated",
    "gravity",
    "gyroscope",
    "gyroscope uncalibrated",
    "linear acceleration",
    "rotation vector",
    "game rotation vector",
    "geomagnetic rotation vector",
    "magnetic field",
    "magnetic field uncalibrated",
    "orientation",
    "proximity",
    "light",
    "pressure",
    "ambient temperature",
    "relative humidity",
    "step counter",
  ];
</script>

<script lang="ts">
  import { ValidationMessage } from "@felte/reporter-svelte";

  export let setFields = (
    path: string,
    value: string[],
    shouldTouch?: boolean
  ) => {};
  export let selected: string[] = [];
  export let updateDirty = () => {};

  const onSelectAll = () => {
    let dirty = sensors.filter((it) => !selected.includes(it)).length > 0;
    setFields("android.sensors", [...sensors], true);
    if (dirty) {
      updateDirty();
    }
  };

  const onClear = () => {
    let dirty = selected.length > 0;
    setFields("android.sensors", []);
    if (dirty) {
      updateDirty();
    }
  };
</script>

<article class="border" aria-labelledby="android-sensors-group-title">
  <details open>
    <summary class="none">
      <div class="row">
        <h6 id="android-sensors-group-title">Android Sensors</h6>
        <div class="max" />
        <i>arrow_drop_down</i>
      </div>
    </summary>

    <ValidationMessage for="android.publisher" let:messages={message}>
      <div class="field label">
        <input type="text" name="android.publisher" id="publisher" />
        <label for="publisher">publisher</label>
        <span class="error">{message || ""}</span>
      </div>
    </ValidationMessage>

    <label class="switch icon">
      <input type="checkbox" name="android.map" aria-label="map pane switch" />
      <span>
        <i>map</i>
      </span>
      <div class="tooltip">map pane</div>
    </label>

    <label class="switch icon">
      <input type="checkbox" name="android.lte" aria-label="lte chart switch" />
      <span>
        <i>signal_cellular_alt</i>
      </span>
      <div class="tooltip">LTE chart</div>
    </label>

    <fieldset class="grid" aria-label="sensor list">
      {#each sensors as sensor}
        <label class="checkbox s12 m6 l3" aria-label={sensor}>
          <input type="checkbox" name="android.sensors" value={sensor} />
          <span>{sensor}</span>
        </label>
      {/each}
    </fieldset>

    <button
      class="round secondary"
      type="button"
      on:click={onSelectAll}
      aria-labelledby="btn-all-label"
    >
      <i>check_box</i>
      <span id="btn-all-label">Select All</span>
    </button>
    <button
      class="round secondary"
      type="button"
      on:click={onClear}
      aria-labelledby="btn-clear-label"
    >
      <i>check_box_outline_blank</i>
      <span id="btn-clear-label">Clear</span>
    </button>
  </details>
</article>

<style>
  .field.label > label {
    color: silver;
  }

  fieldset {
    margin: 0.9375rem;
    border-width: 0;
  }

  .checkbox + .checkbox {
    margin-left: 0;
  }
</style>
