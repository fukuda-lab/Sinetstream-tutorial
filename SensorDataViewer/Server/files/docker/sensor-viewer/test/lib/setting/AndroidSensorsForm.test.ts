import {
  cleanup, fireEvent, getAllByRole, render, screen,
} from '@testing-library/svelte';
import {
  afterEach, describe, expect, it, vi,
} from 'vitest';
import AndroidSensorsForm, {
  sensors as sensorList,
} from '../../../src/lib/setting/AndroidSensorsForm.svelte';

describe('AndroidSensorsForm.svelt', () => {
  afterEach(() => cleanup());

  it('mounts', () => {
    const { container } = render(AndroidSensorsForm);
    expect(container).toBeTruthy();
  });

  it('publisher', () => {
    render(AndroidSensorsForm);
    const item = screen.getByRole('textbox', { name: 'publisher' });
    expect(item).toBeTruthy();
    expect(item.parentElement?.outerHTML).toMatchSnapshot();
    expect(item).toHaveProperty('name', 'android.publisher');
    expect(item).toHaveProperty('type', 'text');
  });

  it('map switch', () => {
    render(AndroidSensorsForm);
    const item = screen.getByRole('checkbox', { name: 'map pane switch' });
    expect(item).toBeTruthy();
    expect(item.parentElement?.outerHTML).toMatchSnapshot();
    expect(item).toHaveProperty('name', 'android.map');
  });

  it('lte switch', () => {
    render(AndroidSensorsForm);
    const item = screen.getByRole('checkbox', { name: 'lte chart switch' });
    expect(item).toBeTruthy();
    expect(item.parentElement?.outerHTML).toMatchSnapshot();
    expect(item).toHaveProperty('name', 'android.lte');
  });

  it('sensors', () => {
    render(AndroidSensorsForm);
    const fieldset = screen.getByRole('group', { name: 'sensor list' });
    const checks = getAllByRole(fieldset, 'checkbox');
    expect(checks).toBeTruthy();
    expect(checks).toHaveLength(18);
    checks.forEach((item) => {
      expect(item).toHaveProperty('name', 'android.sensors');
    });
  });

  it('expansion open', () => {
    render(AndroidSensorsForm);
    const parent = screen.getByRole('article', { name: 'Android Sensors' });
    const item = parent.querySelector('details');
    expect(item?.getAttribute('open')).not.toBeNull();
  });

  describe('select all button', () => {
    it('Nothing is selected.', () => {
      const setFields = vi.fn();
      const updateDirty = vi.fn();
      render(AndroidSensorsForm, { props: { setFields, updateDirty } });
      const item = screen.getByRole('button', { name: 'Select All' });
      expect(item).toBeTruthy();
      fireEvent.click(item);
      expect(setFields).toHaveBeenCalledOnce();
      expect(setFields).toHaveBeenCalledWith('android.sensors', sensorList, true);
      expect(updateDirty).toHaveBeenCalledOnce();
    });

    it('All selected', () => {
      const setFields = vi.fn();
      const updateDirty = vi.fn();
      const selected = [...sensorList];
      render(AndroidSensorsForm, { props: { setFields, updateDirty, selected } });
      const item = screen.getByRole('button', { name: 'Select All' });
      expect(item).toBeTruthy();
      fireEvent.click(item);
      expect(setFields).toHaveBeenCalledOnce();
      expect(setFields).toHaveBeenCalledWith('android.sensors', sensorList, true);
      expect(updateDirty).not.toHaveBeenCalled();
    });
  });

  describe('clear button', () => {
    it('Nothing is selected.', () => {
      const setFields = vi.fn();
      const updateDirty = vi.fn();
      render(AndroidSensorsForm, { props: { setFields, updateDirty } });
      const item = screen.getByRole('button', { name: 'Clear' });
      expect(item).toBeTruthy();
      fireEvent.click(item);
      expect(setFields).toHaveBeenCalledOnce();
      expect(setFields).toHaveBeenCalledWith('android.sensors', []);
      expect(updateDirty).not.toHaveBeenCalled();
    });

    it('All selected', () => {
      const setFields = vi.fn();
      const updateDirty = vi.fn();
      const selected = [...sensorList];
      render(AndroidSensorsForm, { props: { setFields, updateDirty, selected } });
      const item = screen.getByRole('button', { name: 'Clear' });
      expect(item).toBeTruthy();
      fireEvent.click(item);
      expect(setFields).toHaveBeenCalledOnce();
      expect(setFields).toHaveBeenCalledWith('android.sensors', []);
      expect(updateDirty).toHaveBeenCalledOnce();
    });
  });
});
