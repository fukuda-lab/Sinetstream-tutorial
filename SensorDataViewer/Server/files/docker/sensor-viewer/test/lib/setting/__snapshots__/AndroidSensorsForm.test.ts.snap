// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AndroidSensorsForm.svelt > lte switch 1`] = `"<label class=\\"switch icon s-0aZZ0OQmx-Dz\\"><input type=\\"checkbox\\" name=\\"android.lte\\" aria-label=\\"lte chart switch\\" class=\\"s-0aZZ0OQmx-Dz\\"> <span class=\\"s-0aZZ0OQmx-Dz\\"><i class=\\"s-0aZZ0OQmx-Dz\\">signal_cellular_alt</i></span> <div class=\\"tooltip s-0aZZ0OQmx-Dz\\">LTE chart</div></label>"`;

exports[`AndroidSensorsForm.svelt > map switch 1`] = `"<label class=\\"switch icon s-0aZZ0OQmx-Dz\\"><input type=\\"checkbox\\" name=\\"android.map\\" aria-label=\\"map pane switch\\" class=\\"s-0aZZ0OQmx-Dz\\"> <span class=\\"s-0aZZ0OQmx-Dz\\"><i class=\\"s-0aZZ0OQmx-Dz\\">map</i></span> <div class=\\"tooltip s-0aZZ0OQmx-Dz\\">map pane</div></label>"`;

exports[`AndroidSensorsForm.svelt > publisher 1`] = `"<div class=\\"field label s-0aZZ0OQmx-Dz\\"><input type=\\"text\\" name=\\"android.publisher\\" id=\\"publisher\\" class=\\"s-0aZZ0OQmx-Dz\\"> <label for=\\"publisher\\" class=\\"s-0aZZ0OQmx-Dz\\">publisher</label> <span class=\\"error s-0aZZ0OQmx-Dz\\"></span></div>"`;
